### DISCUSSION TITLE
Question about ADSR Sustain Fields

### DISCUSSION THREAD

**Discussion by Meerkov** [vgmtrans/vgmtrans#295] [Music formats]
In PSXSPU.h there are comments to the effect that Sustain is not used, because SF2 and DLS do not support it.

```c++
  // The following are unimplemented in conversion (because DLS and SF2 do not support Sustain Rate)
  uint8_t Sm = (ADSR2 & 0x8000) >> 15;
  uint8_t Sd = (ADSR2 & 0x4000) >> 14;
  uint8_t Sr = (ADSR2 >> 6) & 0x7F;

```

I believe I've run into a few bugs which seem to indicate people noticing that the sound doesn't quite have the right amount of sustain (#138 , #123)

From what I've found on the internet, there seems to be agreement that SF2 and DLS don't support this feature, but there are workarounds... however I'm pretty unfamiliar with a lot of this stuff... and maybe it's already using some work arounds.

What's the status of ADSR support currently? Is this something that needs to be brought up to modern times, or is this still unfixable?

**Comment by Meerkov** [comment]
I was able to find this reference material, which may offer SF2 compliant alternatives to ADSR, but I'm not sure.

http://www.synthfont.com/SFSPEC21.PDF
https://github.com/FluidSynth/fluidsynth/wiki/FluidFeatures

**Comment by loveemu** [comment]
I don't have much knowledge about DLS or PSXSPU implementations, but can comment on a few from my experience implementing ADSR conversions for SF2 and SNES.

The SNES S-DSP has two volume decreasing envelopes, Decay and Sustain (there is also Release, but I won't go into that here). These decrease the volume at different rates. Envelope curves are almost always exponential, with a few exceptions.
On the other hand, I believe that there is only one sustain defined in SoundFont 2.04, and while the terms decay and sustain may appear in SoundFont, they refer to the same volume envelope. To use a SNES analogy, SoundFont is like DR and SL can be fine-tuned, but SR is always infinite. I'm not very familiar with DLS, but perhaps this is the same as SF2.

Therefore, when converting to SoundFont, ADSR has to be expressed as an approximation (except in the case of SL=7). The method I finally adopted was to implement an emulation of the ADSR unit and calculate an approximation based on the hints from it.

1. Emulate ADSR and get hints
   * Get the time it takes for Decay to reach -100dB (DR)
   * Get the volume level at the start of Sustain (SL: in decibels)
   * Get the time it takes for Sustain to reach -100dB when it starts at 0dB (SR)
2. Calculate the decay time to be set for SoundFont by dividing DR and SR proportionally based on SL.
   ```cs
   var SL = SL_dB / -100.0;
   var decay_time = (DR * SL) + (SR * (1.0 - SL));
   ```

Initially, I was thinking of a strategy to focus on either Decay or Sustain depending on how fast Decay is, but I gave up because I thought it would be difficult to set a threshold that would work for all cases.

This method seems to have worked because the SNES Decay, Sustain, and SoundFont Decay were all (almost) exponentially decreasing. If I remember correctly, the PSXSPU has two types of sustain and the volume curve can be selected between linear and exponential envelopes, so I think it is more difficult.

**Reply by Meerkov** [reply]
It looks like there are some sections commented out:
```c++
// We need to compensate the decay and release times to represent them as the time from full vol to -100db
  // where the drop in db is a fixed amount per time unit (SoundFont2 spec for vol envelopes, pg44.)
  //  We assume the psx envelope is using a linear scale wherein envelope_level / 2 == half loudness.
  //  For a linear release mode (Rm == 0), the time to reach half volume is simply half the time to reach 0.
  // Half perceived loudness is -10db. Therefore, time_to_half_vol * 10 == full_time * 5 == the correct SF2 time
  //realADSR->decay_time = LinAmpDecayTimeToLinDBDecayTime(realADSR->decay_time, 0x800);
  //realADSR->sustain_time = LinAmpDecayTimeToLinDBDecayTime(realADSR->sustain_time, 0x800);
  //realADSR->release_time = LinAmpDecayTimeToLinDBDecayTime(realADSR->release_time, 0x800);
  ```
  
  Would this be what you're referring to? It sort of looks like there is an attempt to use sustain when Decay isn't used in the code above these comments, but then there is this stuff which seems to indicate that maybe that approach didn't work right?

**Reply by mikelow** [reply]
Hi, Meerkov. You identified one of the annoying deficiencies of the DLS and SF2 formats. Playstation, CPS2/3, and possibly SNES (don't recall) all have a concept of a sustain rate, but DLS and SF2 do not. For Playstation and CPS2/3, when the decay rate is effectively unused, I substituted in the sustain rate to get the same effect. 

You'll see a comment regarding this and implementation in PSXSPU.h [here](https://github.com/vgmtrans/vgmtrans/blob/d9ce3c8b30da3badeb19608397ffc456849e2b6b/src/main/formats/PSXSPU.h#L291) and in CPSInstr.cpp [here](https://github.com/vgmtrans/vgmtrans/blob/d9ce3c8b30da3badeb19608397ffc456849e2b6b/src/main/formats/CPSInstr.cpp#L438).
